remotecommand: treat unexpected EOF as error to avoid returning success in case of network blip

bimmlerd opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Version 4 (and newer) of the remote command protocol specify that it is expected to get a `metav1.Status` response after termination of the remote process. However, the implementation of the client side of this protocol considered an `EOF` of the error stream (on which this response is expected) to be a successful termination, even if no such `Status` message was received.

This patch changes that behaviour for v4 and up, by letting the (already protocol-specific) error decoder handle the zero length message case. In the v4 error decoder, then, we consider this case an error (as we do with any other unexpected response). The reason I think this change is necessary is that it is possible to observe this EOF in case of a network blip somewhere along the (reasonably long) path from client-go to the CRI.

Notably, this fixes the case where `StreamWithContext` could return successfully (i.e. nil error) even if the program exited with a non-zero exit code, or didn't exit at all. This has caused flakes in cilium's CI in the past, as it heavily relies on this mechanism. The related cilium issue is https://github.com/cilium/cilium/issues/31067.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

Probably worth reviewing commit by commit - I intentionally put the test changes into the first two commits to demonstrate that they pass with the existing implementation. Then, the implementation change in the third can be done without changing tests, increasing confidence. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixes an issue where remotecommand StreamWithContext could succeed even though the underlying transport failed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

bimmlerd committed:
remotecommand: treat EOF as an error for v4+

Version 4 (and newer) of the remote command protocol specify that it is
expected to get a metav1.Status response after termination of the remote
process. However, the client-go implementation of the client side of
this protocol considered an EOF of the error stream (on which this
response is expected) to be a successful termination of
StreamWithContext.

This patch changes that behaviour for v4 and up, by specifying whether
an empty message is to be considered a success.

Notably, this fixes the case where StreamWithContext could return
successfully due to network blips severing the connection, and hence
causing an EOF of the error stream, which has caused flakes in cilium's
CI in the past.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
remotecommand: tests: write status to stream

In v4, the protocol postulates that a metav1.Status is written to the
stream before termination. Some tests didn't do this. Luckily, there is
an existing test server fixture which handles this correctly, so let's
use that to fix the tests.

The main point of this change is to ensure that the tests are correct on
the protocol level, so that the change in the next patch (which makes
unexpected EOFs an error) can pass them without modification.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

aojea commented:
/ok-to-test
@bimmlerd please sign the CLA :)

aojea commented on staging/src/k8s.io/client-go/tools/remotecommand/v4.go:
```diff
@@ -80,6 +81,9 @@ func (p *streamProtocolV4) stream(conn streamCreator) error {
 type errorDecoderV4 struct{}
 
 func (d *errorDecoderV4) decode(message []byte) error {
+	if len(message) == 0 {
+		return fmt.Errorf("error stream protocol error: no message: %w", io.ErrUnexpectedEOF)
+	}
```
this returns an error anyway
```diff
iff --git a/staging/src/k8s.io/client-go/tools/remotecommand/v4_test.go b/staging/src/k8s.io/client-go/tools/remotecommand/v4_test.go
index 217995415ea..6dc14a403a9 100644
--- a/staging/src/k8s.io/client-go/tools/remotecommand/v4_test.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/v4_test.go
@@ -31,6 +31,10 @@ func TestV4ErrorDecoder(t *testing.T) {
        }
 
        for _, test := range []Test{
+               {
+                       message: "",
+                       err:     "error stream protocol error: unexpected end of JSON input in \"\"",
+               },
                {
                        message: "{}",
                        err:     "error stream protocol error: unknown error",
```

```
go test -timeout 30s -run ^TestV4ErrorDecoder$ k8s.io/client-go/tools/remotecommand -v -count 1

=== RUN   TestV4ErrorDecoder
--- PASS: TestV4ErrorDecoder (0.00s)
PASS
ok  	k8s.io/client-go/tools/remotecommand	0.017s
```

aojea commented on staging/src/k8s.io/client-go/tools/remotecommand/v2.go:
```diff
@@ -195,5 +195,8 @@ func (p *streamProtocolV2) stream(conn streamCreator) error {
+	if len(message) > 0 {
+		return fmt.Errorf("error executing remote command: %s", message)
+	}
+	return nil
```
I thought this is the behavior we want to fail, no?

bimmlerd replied:
So I was under the impression that we only started explicit success messages in v4, and this is v2 - if I don't add this case, wouldn't that break v2 compatability?

aojea commented:
we have keepalive frames to detect network hiccups, can the streaming layer detect the abrupt network connection drop and cascade the error instead of relying on the size of the message?

/cc @liggitt @seans3

bimmlerd committed:
proxy: tests: send success status after streaming

A subsequent commit will turn an unexpected EOF into an error, which
would fail these tests. Hence, send the success status to adhere to v4
of the streaming protocol.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd commented:
/retest

bimmlerd commented:
> we have keepalive frames to detect network hiccups, can the streaming layer detect the abrupt network connection drop and cascade the error instead of relying on the size of the message?

@aojea I've traced this a bit further down the stack, and what I've found so far is the following:

1. The layers are as follows: v4 streaming protocol over SPDY over TLS over TCP.
2. On the client, the EOF bubbles up directly from TCP all the way to the streaming protocol- SPDY gets it here https://github.com/moby/spdystream/blob/master/spdy/read.go#L165, from there it bubbles up to the `io.ReadAll` and we hit the default case in `errorstream.go`.
3. Given that we get an EOF from TCP (versus some timeout error), I believe that the server closes the connection, possibly because it doesn't get the keep-alive pings you mention. 
 
Notably, my current repro occurs with repeating cycles of 20s 100% packet loss, 10s no packet loss. The EOF is received only once the 20s are through, ie once TCP is flowing again - hence I guess the server closes the conn without sending an error. It's debatable whether that's another bug, or the root cause.

My repro is essentially a cilium on kind install, where I docker exec onto the control plane node, and run
```
while true; do 
  tc qdisc add dev eth0 root netem loss 100%
  echo "DOWN"; sleep 20 
  tc qdisc del dev eth0 root netem loss 100%
  echo "UP"; sleep 10; 
done
```
once `cilium connectivity test --test client-ingress-knp` has set up and is starting its tests.

bimmlerd commented:
FWIW, I think there's additionally a bug in the upgradeaware proxy code in the API server - see https://github.com/cilium/cilium/issues/31067#issuecomment-1985200305. Do I attempt a fix of that in this PR or open a separate one?

jiahuif commented:
/assign @aojea 
/triage accepted

bimmlerd commented:
@aojea anything I can do to move this forward? I'm hesitant to put up a second PR to fix https://github.com/cilium/cilium/issues/31067#issuecomment-1985200305 if there's some misalignment on whether these fixes are correct/required/wanted at the moment.

liggitt commented on staging/src/k8s.io/client-go/tools/remotecommand/errorstream.go:
```diff
@@ -42,10 +42,8 @@ func watchErrorStream(errorStream io.Reader, d errorStreamDecoder) chan error {
 		switch {
 		case err != nil && err != io.EOF:
 			errorChan <- fmt.Errorf("error reading from error stream: %s", err)
-		case len(message) > 0:
```
keep this case as-is

bimmlerd replied:
done

liggitt commented on staging/src/k8s.io/client-go/tools/remotecommand/v4_test.go:
```diff
@@ -31,6 +31,10 @@ func TestV4ErrorDecoder(t *testing.T) {
 	for _, test := range []Test{
+		{
+			message: "",
+			err:     "error stream protocol error: unexpected end of JSON input in \"\"",
```
this is a really confusing message for v4 to return when the real problem is that no completion status was returned over the error stream... can we improve this?

liggitt commented:
> Version 4 (and newer) of the remote command protocol specify that it is expected to get a `metav1.Status` response after termination of the remote process.

Is it guaranteed we get a status response, or is that only required to convey a non-zero exit code?

bimmlerd commented on staging/src/k8s.io/client-go/tools/remotecommand/errorstream.go:
```diff
@@ -42,10 +42,8 @@ func watchErrorStream(errorStream io.Reader, d errorStreamDecoder) chan error {
-		case len(message) > 0:
-			errorChan <- d.decode(message)
 		default:
-			errorChan <- nil
```
good idea, adopted.

bimmlerd reviewed (commented):
> > Version 4 (and newer) of the remote command protocol specify that it is expected to get a `metav1.Status` response after termination of the remote process.
> 
> Is it guaranteed we get a status response, or is that only required to convey a non-zero exit code?

@liggitt As far as I understand, that's the behaviour of v4. See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/cri/streaming/remotecommand/exec.go#L76-L78 and then the two `writeStatus` implementations https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/cri/streaming/remotecommand/httpstream.go#L426-L447

liggitt commented:
> @liggitt As far as I understand, that's the behaviour of v4. See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/cri/streaming/remotecommand/exec.go#L76-L78 and then the two `writeStatus` implementations https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/cri/streaming/remotecommand/httpstream.go#L426-L447

That's up to the CRI, right? is that _guaranteed_, or just happens to be the behavior of the kubelet and containerd implementations we know of? What would we do if we make this change, then find out a CRI impl was simply closing the connection with no status on successful exit?

bimmlerd commented:
> That's up to the CRI, right? is that _guaranteed_, or just happens to be the behavior of the kubelet and containerd implementations we know of? What would we do if we make this change, then find out a CRI impl was simply closing the connection with no status on successful exit?

@liggitt I understand the concern, I don't want to break the world either. I don't think it is up to the CRI, though. The flow as I understand it: `client -> apiserver -> kubelet -> CRI`

The lines of code linked above are part of kubelet - independent of the specifics of which CRI is used. Kubelet, so far as I understand it, then forwards the request to the actual CRI via its `Executor` interface, which is implemented by `criAdapter` (by turning it into a gRPC RuntimeService `Exec` call). Therefore, since the kubelet guarantees to write a `metav1.Status`, I think there's no way the CRI implementation can change that. 

I think it's somewhat difficult to follow the flow in code though and hence I might be missing the architectural picture.
